model/riscv_inst_retire.sail

```diff
@@ -0,0 +1,30 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+/* Reasons an instruction retire might fail or be incomplete. */
+
+union Retire_Failure = {
+  // standard reasons
+  Illegal_Instruction            : unit,
+  Wait_For_Interrupt             : unit,
+  Trap                           : (Privilege, ctl_result, xlenbits),
+  Memory_Exception               : (virtaddr, ExceptionType),
+
+  // reasons from external extensions
+  Ext_CSR_Check_Failure          : unit,
+  Ext_ControlAddr_Check_Failure  : ext_control_addr_error,
+  Ext_DataAddr_Check_Failure     : ext_data_addr_error,
+  Ext_XRET_Priv_Failure          : unit,
+}
+
+// To ease the introduction of `Retire_Failure`, this global
+// definition is temporarily used instead of the previous
+// RETIRE_SUCCESS enum value so that we don't have to immediately
+// update all the places RETIRE_SUCCESS was used.
+
+let RETIRE_SUCCESS : ExecutionResult(Retire_Failure) = RETIRE_OK()
```

Comment:
Should/can this be polymorphic and down (up?) in `riscv_types.sail`, where `ExecutionResult` and `RETIRE_OK` are defined?  Is the monomorphism of the not-used `union` arm convenient for type inference at use sites?

Reply:
I should have added
```
let RETIRE_SUCCESS = RETIRE_OK()
```
gives a type error.

Reply:
Can top-level `let`s take a type annotation?
```
let RETIRE_SUCCESS : forall 'a . ExecutionResult('a) = RETIRE_OK()
```

Reply:
No, there is a syntax error at `forall`.